osquery/tables/system/windows/certificates.cpp

```diff
@@ -541,12 +542,27 @@ Status expandEnvironmentVariables(const std::string& src, std::string& dest) {
   auto ret = ExpandEnvironmentStringsW(srcW, buf.data(), expandedSize);
   if (ret == 0) {
     return Status::failure("Environment variable expansion failed");
+  } else if (ret != expandedSize) {
+    return Status::failure("Partial data written");
   }
 
   dest = wstringToString(buf.data());
   return Status::success();
 }
 
+Status expandEnvironmentVariables(const std::string& src, std::string& dest) {
+  const int MAX_RETRY = 100;
```

Comment:
So my 2c would be to get rid of this wrapper and just attempt the call one more time. This does seem like a good assurance to get the data, but you’re having success for the most part with the one extra call I’d say just use that, as this feels a bit excessive to me. What do you think?

Reply:
I think that's a good point, but retrying just one more time does make me a little uneasy... maybe it's just a peculiarity of my machine that it always succeeds on the second try? Maybe 1 out of 500 tries requires more than 2 tries, and I just have never observed that case? Maybe on most other machines, it's something like 3-5?

The risk here is inconsistent and mysterious looking behavior from the user perspective, and I think it's worth going a little "overboard" on our end if it will make the table "substantially" more reliable for end users. I realize that the retry number is totally arbitrary, but I do think using a retry loop with some "large-ish" iteration number will decrease the likelihood of the table being unreliable to a negligible point. I'm not sure the same is true for adding 1 more retry.

In this case I'd say the benefits of extra safety outweigh the maintainability/readability downsides. You have more experience with this than I though :) I'm curious how you've weighed this decision in the past. (I only thought to do the fixed iteration loop because of your previous review comments on #5640)

Reply:
The reason it doesn't work is that a ptr to a temporary is saved in the wrapper:

```
  auto srcW = stringToWstring(src).c_str();
  auto expandedSize = ExpandEnvironmentStringsW(srcW, nullptr, 0);
  if (expandedSize == 0) {
    return Status::failure("Unable to get expanded size");
  }
```

Reply:
Oof, thanks Alessandro. I fixed that bug, and now it seems like the retry loop is no longer necessary. The docs still don't seem to guarantee that incomplete writes _can't_ happen, so I added a little check just to explicitly bail out, but have removed the loop.

https://docs.microsoft.com/en-us/windows/win32/api/processenv/nf-processenv-expandenvironmentstringsw